packages/http/src/types.ts

```diff
@@ -35,7 +35,20 @@ export type HttpAuth =
   | BearerAuth
   | ApiKeyAuth<ApiKeyLocation, string>
   | Oauth2Auth<OAuth2Flow[]>
-  | OpenIDConnectAuth;
+  | OpenIDConnectAuth
+  | NoAuth;
+
+export interface HttpAuthReference {
+  /*
+   * Unique id of the authentication scheme among all schemes.
+   * `undefined` means lack of auth
+   */
+  id?: string;
+  /**
+   * OAuth2 flows. Defined only when `id` is referencing an `OAuth2Auth`
+   */
+  oauth2Flows?: OAuth2Flow[];
```

Comment:
wondering here if we can't be exposing the `scopes` directly instead. It feels like if there is different flows used for an operation then it should be pointing to an different `HttpAuth`

Reply:
Maybe if we wanted to be safe and accurate we could make this type a union of 
```ts
type HttpAuthReference = SimpleRef | OAuth2Ref;
```

Reply:
Came up with these types:
```typescript
export type HttpAuthRef = AnyHttpAuthRef | OAuth2HttpAuthRef | NoHttpAuthRef;

export interface AnyHttpAuthRef {
  readonly kind: "any";
  readonly auth: HttpAuth;
}

export interface NoHttpAuthRef {
  readonly kind: "noAuth";
  readonly auth: NoAuth;
}

/* Holder of this reference needs only a `scopes` subset of all scopes defined at `auth` */
export interface OAuth2HttpAuthRef {
  readonly kind: "oauth2";
  readonly auth: Oauth2Auth<OAuth2Flow[]>;
  readonly scopes: string[];
}
```

Reply:
I think that's a good option. I would maybe instead for the `AnyHttpAuthRef` just have an open string and make sure the `kind` are the same as the auth that it will reference

Reply:
This would make Oauth2HttpAuthRef and AnyHttpAuthRef indistinguishable in the following switch statement:
```
        switch (httpAuthRef.kind) {
          case "noAuth":
            // should emit "{}" as a security option https://github.com/OAI/OpenAPI-Specification/issues/14#issuecomment-297457320
            continue;
          case "oauth2":
            securityOption[httpAuthRef.auth.id] = httpAuthRef.scopes;
            continue;
         ...
```

The second branch would match `AnyHttpAuthRef | OAuth2HttpAuthRef`, which doesn't have a `scopes` property

Reply:
True, but what if you have 1 ref kind per auth kind? Just thinking that if we ever need properties on those ref it would be a breaking change to add that extra kind

Reply:
Hhm though I guess the same could be said about adding a new auth kind. Maybe it’s fine then